CIP-XXXX/README.md

```diff
@@ -0,0 +1,173 @@
+
+```ts
+type KeyHash = String
+```
+
+#### ScriptRequirements
+
+```ts
+type ScriptRequirementsCode = {
+    Signer: 1,
+    Before: 2,
+    After: 3,
+}
+type ScriptRequirement = {
+    code: ScriptRequirementsCode,
+    value: KeyHash|number,
+}
+```
+
+### Aditional Error Types
+
+#### CompletedTxError
+
+```ts
+CompletedTxErrorCode = {
+	NotFound: 1,
+	NotReady: 2
+}
+```
+
+* NotFound - The transaction with the given id was not found.
+* NotReady - The transaction with the given id is not ready yet. 
+
+## Motivation: why is this CIP necessary?
+In order to facilitate future dApp development, we will need a way for dApps to communicate with multisig wallets. Given the unique complexities of native script based addresses, special provisions need to be made to make the connector compatible with them.
+
+Specifically, apps building transactions need to be able to get the following information from the wallet:
+
+- Script descriptor
+- Script Requirements list
+- Collateral donator (since native script based addresses cannot provide collateral for transactions)
+
+Additionally, apps need to be able to submit a transaction to the wallet for signing in an asynchronous manner, as gathering of signatures can take a long time and each wallet provider will have its own way of handling this process.
+
+Finally, the signTx() and signData() endpoints will have to be disabled when using this extension since they are not compatible with native script based addresses.
+
+## Rationale for the required data
+- Script descriptor: Any transaction consuming a UTxO from a native script based address must attach the corresponding script.
+- Script Requirements list: 
+	-- dApps need to know the number of signers to calculate the fees correctly. 
+	-- Including the correct "required signers" brings UX improvements. 
+	-- dApps need to know if the script has any before-after requirements to set the correct validBefore and TTL values.
+- Collateral donator Native script based addresses cannot provide collateral for Plutus transactions.
+
+### Additional API Endpoints
+
+#### api.getCollateralAddress(): Promise\<Address>
+
+For Plutus V2 and later, partial collateral is supported. This function returns an address that can be used to add collateral to a transaction. The address returned must be owned by one of the signers in the list of signers returned by `api.getScriptRequirements()`. 
+
+dApp developers can choose to use this address to add collateral to a transaction, or they can choose to use the `api.getCollateral()` function to get a list of UTXOs that can be used as collateral. If the dApp chooses to use this address, they must ensure that the address is not used for any other purpose, as the wallet may be using it to track collateral, and that the collateral return address is the same one.
+
+#### api.getScriptRequirements: Promise\<ScriptRequirement[]>
```

Comment:
As mentioned there are 3 reasons to set this structure and not return a simple int for the number of signatures.
- Script Requirements list:
	-- dApps need to know the number of signers to calculate the fees correctly.
	-- Including the correct "required signers" brings UX improvements.
	-- dApps need to know if the script has any before-after requirements to set the correct validBefore and TTL values.

Reply:
In your example wallet implementers are responsible to create the UI/UX where the user choses if the signers are (A+B) or C. 

So the output of the connector will be (A+B). 

Or if the user specified C the connector will return only (C)

Reply:
> -- Including the correct "required signers" brings UX improvements.
> -- dApps need to know if the script has any before-after requirements to set the correct validBefore and TTL values.

I'm not sure about the value of those - user could see signing info and status in the wallet. Also if the dapp was multi-sig specific, then it would probably be able to parse the script from `getScript`. I'm probably missing some dapp use case(s) here, do you have something specific in mind?

> In your example wallet implementers are responsible to create the UI/UX where the user choses if the signers are (A+B) or C.

That sounds reasonable, but then this method seems to be misleading - it's not really the script requirements, more like constraints or a plan for signing the next transaction :thinking:

Reply:
> I'm not sure about the value of those - user could see signing info and status in the wallet. Also if the dapp was multi-sig specific, then it would probably be able to parse the script from getScript. I'm probably missing some dapp use case(s) here, do you have something specific in mind?

you said it your self, native script are not flat. Before and After conditions could exist but not be used to validate a transaction if alternative exits 

example vesting: (A or (B and AFTER x)), 'AFTER x' is only required if A is not one of the signers. the connector must return the exact list or condition to validate a single Tx.

>That sounds reasonable, but then this method seems to be misleading - it's not really the script requirements, more like constraints or a plan for signing the next transaction 🤔

I am open to modifications that will make the use of this endpoint more intuitive.

Reply:
I have a counter example for UX: consider a wallet with a script requiring signature of `A or B`. Then,
- if this API didn't require providing specific signers, then wallet could just say that it will sign with 1 signature
- if this API requires it, wallet has to add an additional step for the user to select a specific signer for the next transaction. And it can't really select a random one, since the intent specified in this CIP is for 'dapp ux' - so it has to assume that dapp would show the signer's address somewhere.

So if you consider the overall UX (dapp + wallet), not just the dapp, then I'd argue it makes it worse in this case.

Reply:
No because including the correct signers in the "Required signers" of the TX allows the wallet to guide the user to find the missing signatures much easier. If you provide a simple int you are left with a user that has no idea who has to sigh what. 

And UX design does not require you ask the user every time.  It can even be done automatically for the initial value but give the user the option to change it in the settings. A proper comminication of dApp-wallet should contain the exact precise information.

Reply:
> No because including the correct signers in the "Required signers" of the TX allows the wallet to guide the user

Wallet already knows the signing rules based on the script, I don't think the dapp should decide which specific signers to require.

>  And UX design does not require you ask the user every time. It can even be done automatically for the initial value but give the user the option to change it in the settings.

The point I was trying to make is that with current API wallet **shouldn't** decide it automatically due to this point in the cip:
> -- Including the correct "required signers" brings UX improvements.

In the `A or B` example, if the wallet automatically choose `A`, but user signs it with `B`, then there's a mismatch between what the dapp potentially shows to the user and the actual signature. Even worse if the dapp sets `requiredSignatures` to `A` - user then can't even sign it with `B`.